### PR TITLE
WIP adding more flexible pricing

### PR DESCRIPTION
**Pull request by transitive-bullshit**
This PR adds a significant degree of flexibility to our pricing model.

The goal is to support more of the common SaaS pricing models that [Stripe supports](https://stripe.com/docs/billing/subscriptions/examples) while keeping the complexity and default pricing configuration as simple as possible.

Here are some of the new use cases that we'll support:

- [x] Add support for more than 2 pricing plans
- [x] Make metered billing optional for plans (leave off `plan.requests` to just charge a base amount or set `plan.requests.usageType = 'licensed'`)
- [x] Add support for tiered pricing plans (`plan.requests.billingScheme = 'tiered'` and `plan.requests.tiers = [ ... ]`)
- [x] Allow rate limit interval to be specified as a human-readable string instead of just as a number of seconds (`1h`, `60m`, `1 day`, [etc](https://github.com/zeit/ms))
- [x] Allow services to disable or use custom rate limits (`service.rateLimit = null` or `service.rateLimit = { requestsInterval: '1h', requestsMaxPerInterval: 1000 }`)
- [x] Allow services to never report usage to stripe (`service.reportUsage = false`)
- [x] Allow services to override their config based on the active pricing plan
  - disable a service on the "free" plan `service.pricingPlanConfig = { free: { enabled: false } }` (calls to this service will result in 403 if the caller is using the "free" plan)
  - disable rate limiting for a service on the "unlimited" plan `service.pricingPlanConfig = { unlimited: { rateLimit: null } }`
  - disable reporting usage to stripe for a service on the "pro" plan `service.pricingPlanConfig = { pro: { reportUsage: false } }`

pricing-a

### CODE REVIEW

File: packages/saasify-cli/lib/schemas/config.ts [hunk 1]
```diff
@@ -4,34 +4,102 @@
...
-  timeout?: number = 0
   examples?: Example[]
   config?: object
   GET?: boolean
   POST?: boolean
   headers?: object
   immutable?: boolean
+
+  // whether to report calls to this service (for metered usage)
+  reportUsage?: boolean = true
```
Comment (chrisvxd) [inline packages/saasify-cli/lib/schemas/config.ts]: Don't understand what this is for

File: packages/saasify-cli/lib/schemas/config.ts [hunk 2]
```diff
@@ -4,34 +4,102 @@
...
+  // UI-only list of features to display for this plan (supports Markdown)
+  features?: string[]
+
+  amount?: number = 0
+  requests?: PricingPlanMetric
+
+  rateLimit?: RateLimit
+}
+
+class PricingPlanMetric {
```
Comment (chrisvxd) [inline packages/saasify-cli/lib/schemas/config.ts]: Should this be at the root level rather than the plan level?

**Comment by transitive-bullshit** [comment]
This is how you would do OG Impact:

```
pricingPlans:
  - name: Free
    slug: free
    desc: Free tier
    auth: false
    features:
      - Unlimited images
      - Template gallery (watermarked)
    rateLimit: null
  - name: Starter
    slug: starter
    desc: For startups
    auth: true
    features:
      - Unlimited images
      - Template gallery (unwatermarked)
      - Priority support
    amount: 999
  - name: Business
    slug: starter
    desc: For businesses
    auth: true
    features:
      - Unlimited images
      - Template gallery (unwatermarked)
      - Create your own templates
      - Priority support
    amount: 2999
```

Note that since you're not charging per request ever, you can just leave off the `requests` stuff entirely.

**Comment by chrisvxd** [comment]
I don't think I said, but this is looking really awesome btw. I think this removes all of the limitations I felt with the existing pricing model, and should unlock a lot of opportunities. It's exciting you did it all in one PR.

**Comment by transitive-bullshit** [comment]
Update: this is mainly waiting on the backend side of things and we'll want to do a heavy round of testing before pushing any updates to the pricing stuff on prod.

**Comment by transitive-bullshit** [comment]
Some examples of pricing in the SaaS client with the `waves` theme:

We can condense things if none of the pricing plans use metered billing or rate limiting.

pricing-a

If they do use these features, then we'll show the expanded pricing details.

pricing-b

**Comment by transitive-bullshit** [comment]
Aaaand an example of the default metered billing if no pricing plans are specified:

pricing-c

**Comment by chrisvxd** [comment]
@transitive-bullshit Is the above config for OG IMPACT still valid? If not, mind sharing an up to date one?

File: packages/react-saasify/src/lib/upgrade.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export function getUpgradeLink({ auth, deployment }) {
+  const { pricingPlans } = deployment
+
+  if (auth.isAuthenticated) {
+    if (pricingPlans.length === 2) {
+      return `/chckout?plan=${pricingPlans[1].slug}`
```
Comment (chrisvxd) [inline packages/react-saasify/src/lib/upgrade.js]: Typo in url? `chckout`

**Comment by transitive-bullshit** [comment]
> Is the above config for OG IMPACT still valid? If not, mind sharing an up to date one?

```json
{
  "name": "test-pricing-basic",
  "services": [
    {
      "src": "index.ts",
      "immutable": true
    }
  ],
  "pricingPlans": [
    {
      "name": "Free",
      "amount": 0,
      "rateLimit": null,
      "features": [
        "Unlimited images **with watermark**",
        "No backend generation",
        "No custom templates`"
      ]
    },
    {
      "name": "Starter",
      "amount": 999,
      "features": [
        "Unlimited images **without watermark**",
        "Unlimited backend generation",
        "No custom templates"
      ]
    },
    {
      "name": "Pro",
      "amount": 2999,
      "features": [
        "Unlimited images **without watermark**",
        "Unlimited backend generation",
        "Unlimited custom templates"
      ]
    }
  ]
}
```

File: packages/react-saasify/src/lib/upgrade.js (hunk 3 above)
Comment (transitive-bullshit) [reply packages/react-saasify/src/lib/upgrade.js]: good catch 😄

**Comment by chrisvxd** [comment]
Okay that looks solid.

It looks like you're keeping the plan slug and you still need to configure the plans on the service, rather than configuring services on the plan? My preference is still the latter, but it's not a huge deal at this point.

Re the UI consistency - definitely good for now, especially since we haven't got any use cases for mixing pricing models yet.

**Comment by chrisvxd** [comment]
One last thing actually - I think the "Free. Unlimited. Pay as you grow" tag line should probably be removed (or defined by the user) since it doesn't make sense most of the time now.

**Comment by transitive-bullshit** [comment]
## Release 🚀 

The newest pricing hotness is published & deployed in prod.

Fixed a few minor issues after things were live. Afaict everything is working properly now.

As far as testing goes, here's what i've done so far:

### Automated

- ran our prod e2e tests
- ran all of our unit tests

### Manual

- subscribing to imagemin, making some authenticated api calls, ensuring their usage is being tracked.
- vetted puppet master pricing (some minor okta-related css issues in a few places).
- made sure my existing subscription to synopsis was still recognized and handled properly.

**Pull request merged**